### PR TITLE
Improve external quoting logic

### LINKED ISSUE #3329: Tilde in Quotes With External Command Not Working
**Describe the bug**
When I use a path starting from home with tilde and wrapped into quote, the external command does not find the file.

**To Reproduce**
1. Create a text file in your home, e.g. `~/test.txt`
2. Print the file with `cat "~/text.txt"` (quotes are necessary to reproduce)
3. `cat` complains that it cannot find `'~/test.txt'`

**Expected behavior**

`cat` should print the content of `~/text.txt`

**Configuration (please complete the following information):**

Run `version | pivot` and paste the output to show OS, features, etc.

```
────┬────────────────────┬──────────────────────────────────────────────────────────────────────────────────────────────────────
 #  │      Column0       │                                               Column1                                                
────┼────────────────────┼──────────────────────────────────────────────────────────────────────────────────────────────────────
  0 │ version            │ 0.29.0                                                                                               
  1 │ branch             │ master                                                                                               
  2 │ short_commit       │ ad539cff                                                                                             
  3 │ commit_hash        │ ad539cff8c4ae36b7ce89ebbdb38c9d2277d51f8                                                             
  4 │ commit_date        │ 2021-03-31 01:10:45                                                                                  
  5 │ build_os           │ linux-x86_64                                                                                         
  6 │ rust_version       │ rustc 1.48.0 (7eac88abb 2020-11-16)                                                                  
  7 │ rust_channel       │ stable-x86_64-unknown-linux-gnu (default)                                                            
  8 │ cargo_version      │ cargo 1.48.0 (65cbdd2dc 2020-10-14)                                                                  
  9 │ pkg_version        │ 0.29.0                                                                                               
 10 │ build_time         │ 2021-04-02 15:20:58                                                                                  
 11 │ build_rust_channel │ release                                                                                              
 12 │ features           │ clipboard-cli, ctrlc, default, directories, dirs, ichwh, ptree, rustyline, term, trash, uuid, which, 
    │                    │ zip                                                                                                  
────┴────────────────────┴──────────────────────────────────────────────────────────────────────────────────────────────────────
```

The main branch's version does not work either:

```
────┬────────────────────┬───────────────────────────────────────────────────────────────────────
 #  │      Column0       │                                Column1                                
────┼────────────────────┼───────────────────────────────────────────────────────────────────────
  0 │ version            │ 0.29.2                                                                
  1 │ branch             │ main                                                                  
  2 │ short_commit       │ 033df945                                                              
  3 │ commit_hash        │ 033df9457b32f2e52447e5068cc8e13f577eabd5                              
  4 │ commit_date        │ 2021-04-18 20:40:29                                                   
  5 │ build_os           │ linux-x86_64                                                          
  6 │ rust_version       │ rustc 1.51.0 (2fd73fabe 2021-03-23)                                   
  7 │ rust_channel       │ stable-x86_64-unknown-linux-gnu (default)                             
  8 │ cargo_version      │ cargo 1.51.0 (43b129a20 2021-03-16)                                   
  9 │ pkg_version        │ 0.29.2                                                                
 10 │ build_time         │ 2021-04-18 20:44:30                                                   
 11 │ build_rust_channel │ debug                                                                 
 12 │ features           │ ctrlc, default, directories, dirs, ichwh, rustyline, term, which, zip 
────┴────────────────────┴───────────────────────────────────────────────────────────────────────
```

### PR DESCRIPTION
**Pull request by jonathandturner**
This PR:
* Improves the quoting logic for external arguments
* Adds a new testbin called "meow" which does a `cat`
* Adds a bit of additional escaping logic

fixes #3329

### CODE REVIEW

File: tests/shell/pipeline/commands/external.rs [hunk 1]
```diff
@@ -217,6 +218,27 @@ mod external_words {
 
         assert_eq!(actual.out, "test \"things\"");
     }
+
+    #[test]
+    fn external_arg_with_quotes() {
+        Playground::setup("external_arg_with_quotes", |dirs, sandbox| {
+            sandbox.with_files(vec![FileWithContent(
+                "sample.toml",
```
Comment (schrieveslaach) [inline tests/shell/pipeline/commands/external.rs:226]: @jonathandturner, I'm wondering what happens if the path contains spaces? Or do you tackle this case in another issue?

Comment (jonathandturner) [reply tests/shell/pipeline/commands/external.rs:226]: Paths with spaces should work. I've done a few completion updates, which includes improving how we handle paths with spaces.

Are you seeing any issues with the latest PRs?

Comment (schrieveslaach) [reply tests/shell/pipeline/commands/external.rs:226]: My thought was that the test could reflect that spaces are supported. Like that:

```suggestion
                "a sample file.toml",
```

So, I don't see any trouble just an slight test code improvement.

Comment (jonathandturner) [reply tests/shell/pipeline/commands/external.rs:226]: @schrieveslaach if you're interested, we'd love to have more tests to help cover cases like that.

Comment (schrieveslaach) [reply tests/shell/pipeline/commands/external.rs:226]: I'll take a look. 👍

Comment (schrieveslaach) [reply tests/shell/pipeline/commands/external.rs:226]: @jonathandturner, I created #3596, do you mind and have a look at it?

**Pull request merged**